presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeConstantValuesUnderUnion.java

```diff
@@ -0,0 +1,128 @@
+{
+    private static final Capture<List<PlanNode>> CHILDREN = newCapture();
+    private static final Pattern<UnionNode> PATTERN = union()
+            .with(sources().capturedAs(CHILDREN));
+
+    @Override
+    public Pattern<UnionNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeUnionOverValues(session);
+    }
+
+    @Override
+    public Result apply(UnionNode node, Captures captures, Context context)
+    {
+        List<PlanNode> values = captures.get(CHILDREN);
+        //Return if not union over ValuesNode
+        if (!(values.stream().map(x -> context.getLookup().resolve(x)).allMatch(x -> x instanceof ValuesNode))) {
+            return Result.empty();
+        }
+
+        Map<VariableReferenceExpression, List<RowExpression>> inputVariablesToValuesMap = getInputVariablesToValuesMap(context, values);
+
+        List<List<RowExpression>> finalValues = generateRowsFromValues(node, context, values, inputVariablesToValuesMap);
+        return Result.ofPlanNode(
+                  new ValuesNode(
+                  node.getId(),
+                  node.getOutputVariables(),
+                  finalValues));
+    }
+
+    private List<List<RowExpression>> generateRowsFromValues(UnionNode node, Context context, List<PlanNode> values, Map<VariableReferenceExpression, List<RowExpression>> columnToValuesMap)
+    {
+        Map<VariableReferenceExpression, List<RowExpression>> outputVarToValuesMap = getOutputVariablesToValuesMap(node, columnToValuesMap);
+        List<Iterator> iterators = node.getOutputVariables().stream().map(x -> outputVarToValuesMap.get(x).iterator()).collect(Collectors.toList());
+
+        int totalRows = values.stream().map(x -> (ValuesNode) context.getLookup().resolve(x)).mapToInt(x -> x.getRows().size()).sum();
+        List<List<RowExpression>> finalValues = new ArrayList<>(totalRows);
+
+        for (int i = 0; i < totalRows; i++) {
+            List<RowExpression> row = new ArrayList<>();
+            for (Iterator<RowExpression> iterator : iterators) {
+                row.add(iterator.next());
+            }
+            finalValues.add(row);
+        }
+        return finalValues;
+    }
+
+    private Map<VariableReferenceExpression, List<RowExpression>> getOutputVariablesToValuesMap(UnionNode node, Map<VariableReferenceExpression, List<RowExpression>> columnToValuesMap)
+    {
+        Map<VariableReferenceExpression, List<RowExpression>> outputVarToValuesMap = new HashMap<>();
+        for (Map.Entry<VariableReferenceExpression, List<VariableReferenceExpression>> entry : node.getVariableMapping().entrySet()) {
+            VariableReferenceExpression outputVar = entry.getKey();
+            for (VariableReferenceExpression input : entry.getValue()) {
+                if (outputVarToValuesMap.get(outputVar) == null) {
+                    outputVarToValuesMap.put(outputVar, new ArrayList<>());
```

Comment:
You can use `ImmutableList.Builder` instead of `ArrayList`. I think you can actually also change the `HashMap` to `ImmutableMap` as well.

Reply:
fixed for final values returned